docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+instantiated by the implicit search. They are visible in the rest of
+the case.
+
+In general, a type name in an implicit condition is treated as a bound
+variable if it is prefixed by `type`.
+
+_Aside:_ We should introduce the same convention for types in normal
+patterns. Currently type identifiers in patterns are treated as bound
+if they start with a lower-case letter and as free otherwise. This
+mirrors the convention for names, but is not very visible. Requiring
+`type` prefixes instead would make programs clearer.
+
+In last clause of the code above, the macro interpreter would perform
+an implicit search for the type `N =:= S[type N1]` where `N` is the
+actual type passed to `toInt` and `N1` is a type variable bound by the
+search.  If the search succeeds, it returns one plus the result of
+evaluating `toInt[N1]`. If the search fails, it results in compilation
+failure.
+
+The result of `toInt[S[S[Z]]]` would hence be the integer `1 + (1 +
+0)`, which can be constant-folded to `2`. Alternatvely, we could drop
+the use of quotes and rely on lifting to go from `Int` to
+`Expr[Int]`. So the following would also be correct and would return a
+constant without relying on constant folding:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =: Z => 0
+      case N =:= S[type N1] => 1 + toInt[N1]
+    }
+
+### Example: HLists
+
+Consider the standard definition of `HList`s:
+
+    class HList
+
+    case class HCons[X, Xs <: HList](hd: X, tl: Xs) extends HList
+    case object HNil extends HList
+
+A type-level length function on HLists can be written as follows:
+
+    type Length[Xs <: HList] <: Nat = {
+      case Xs =:= HNil => Z
+      case Xs =:= HCons[type X, type Xs1] => S[Length[Xs1]]
+    }
+
+If we don't want to introduce the type variable `X`, which is unused, we could
+also formulate the type pattern with `<:<` instead of `=:=`:
+
+    type Length[Xs <: HList] <: Nat = {
+      case Xs <:< HNil => Z
+      case Xs <:< HCons[_, type Xs1] => S[Length[Xs1]]
+    }
+
+Here's a `Concat` type with associated `concat` method:
+
+    type Concat[Xs <: HList, Ys <: HList] <: HList = {
+      case Xs =:= HNil => Ys
+      case Xs =:= HCons[type X, type Xs1] => HCons[X, Concat[Xs1, Ys]]
+    }
+
+    inline def concat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys): Concat[Xs, Ys] = ~ {
+      case Xs =:= HNil => 'ys
+      case Xs =:= HCons[type X, type Xs1] => 'HCons(xs.hd, concat(xs.tl, ys))
+    }
+
+### Typechecking Query Conditionals
+
+Why does the last case of `concat` typecheck, given that `xs` is of
+type `Xs`, which does not have `hd` or `tl` fields? Roughly, it
+typechecks because `=:=` (and `<:<`) will inherit from
+`ImplicitConverter` so in the presence of an implicit search result of
+`T =:= U` or `T <:< U` we also get an implicit conversion from `T` to
+`U`. But we still need to explain why a search for `Xs =:= HCons[type X, type Xs1]`
+will succeed in the right hand side of the case. What happens in detail is this:
```

Comment:
Let me assume `concat` is a `def` not an `inline def` (see above).

We also need to make sure that the recursive call to `concat` typechecks — I can imagine ways in which it’d work, but one needs to pick a typechecking algorithm.

The sketched process does not seem to use type refinement — in particular, it seems that in this branch we don’t know that `Xs` is equal to `HCons[X, Xs1]`, only that they’re convertible both ways. But then, the type of the RHS is `HCons[X, Concat[Xs1, Ys]]`, while the desired return type is `Concat[Xs, Ys]` — without unifying `Xs` with `HCons[X, Xs1]`, why would the two coincide? I now see `concat` is here a query conditional, but see above for why that might not work well enough.

Also, will the code still typecheck if I write the following?

```scala
    inline def concat[Xs <: HList, Ys <: HList](xs: Xs, ys: Ys): Concat[Xs, Ys] = ~ {
      case Xs <:< HNil => 'ys
      case Xs <:< HCons[_, type Xs1] => 'HCons(xs.hd, concat(xs.tl, ys))
    }
```

Reply:
Good question. I would assume yes, but without coming up with a detailed algorithm I would not know for sure.

Reply:
This is a really important problem in practice, and not just in the context of metaprogramming – the type-checker does not leverage `<:<` evidence to refine the set of known subtyping relations.

One way I've found that works in Dotty (but not scalac) to help work around this problem is to define `<:<` as:

```scala
abstract class <:< [-A,+B] { type Ev >: A <: B }
```

Then we can use explicit type ascriptions to lead the type from one side to the other, as in:

```scala
def foo[S,T](xs: List[S])(implicit ev: S <:< T): List[T] = (xs: List[ev.Ev])
```

Reply:
IIRC the Scala’17 paper on algorithmic DOT describes the very need for such ascriptions.

Though in Scala you can *call* `<:<` to use it as a coercion function, though IIRC Scalac (used to?) produce for this an actual method call (which JIT can probably inline).

Reply:
> IIRC the Scala’17 paper on algorithmic DOT describes the very need for such ascriptions.

That's right! I was probably inspired by that talk when I found the trick.

> Though in Scala you can call <:< to use it as a coercion function

Note that this is not sufficient by any account. In my example, it means we have to rebuild a new list `xs.map(ev)` which is purely accidental runtime overhead (not present in `foo` above). In some cases, it's not even possible to rebuild the value this way. 

An even harder case is to use `S` in type argument position where a bound `<: T` is required. To make that one work, I think we'd need to use `ev.Ev` instead of `S` but also make `ev.Ev` equivalent to `S`, as in:

```scala
abstract class <:< [A,+B] { type Ev >: A <: B & A }
```

Reply:
(Should we move this topic to a separate issue?)

> Note that this is not sufficient by any account.

Good point, so I suspect I’m in favor of adding `Ev`. I think scalaz/cats have replacements for `<:<` (`Liskov`?), implemented with casts. Similarly, Haskell added the whole `Coercible` typeclass and friends, to play the role of `=:=` while solving this problem. `Coercible` (and *roles*) seem overly complicated but at least can address similar problems.

> An even harder case [...]

Argh. I think you’re right, but I might rewrite the call site to take `S <:< T` instead. But I’m not sure that’s satisfactory, and it currently adds overhead (not sure if Dotty phantom types can help).

I and @sstucki had reasons to consider some *built-in* `<:<` for DOT that the typechecker understands as subtyping evidence, but we never discussed exposing it to users. I’d prefer an encoding such as what you show, though the asymmetry there bugs me (and would be trouble when using `T` for a type variable with bound `>: S`.
What about
```scala
abstract class <:<[-A, +B <: A] {
  type Ev >: A <: B
  type EvB >: A <: A & B /* = B */
  type EvA >: A | B <: B /* = A, sort-of-dual to the above */
}
```
? I’ve also thought about just having `abstract class <:<[-A, +B <: A]` but it has no type members, and type members don’t have declaration-site variance annotations.

Reply:
> (Should we move this topic to a separate issue?)

I would, but @odersky told us at POPL (jokingly I believe) that if we want to open a new issue, we have to close two old ones first 😅 
Anyway, it's probably easier to ask him directly.

It's actually pretty cool that in Dotty you don't need unsafe casts to achieve the type coercion. It does flow naturally from the theory, where objects with bounded abstract type members can truly be witnesses of subtyping relations. AFAIK this constituted one of the difficult parts of proving the core calculus sound, so we better make use of it 😄 

I like the symmetry of your alternative, but then `Ev` seems really redundant, as a type floating in between without providing anything useful. So I'd remove it:

```scala
abstract class <:<[-A, +B <: A] {
  type ConstrainedB >: A <: A & B // = B with constraints
  type ConstrainedA >: A | B <: B // = A with constraints
}
```

I think the next logical step would be to for the type checker to automatically leverage subtyping relations found as members of values in the current scope (maybe only consider those from implicit values?), though as explained in the Scala'17 talk IIRC it's impossible to make this approach complete in general.

> I’ve also thought about just having abstract class `<:<[-A, +B <: A]`

I've tried that too, but it doesn't work with how subtyping knowledge is propagated:

```scala
def foo[A,B](x: A <:< B) = x
                      ^
                      Type argument B does not conform to upper bound A
```

Reply:
> [On removing Ev]
Sure.

> [automatically leverage subtype relations]

I should look at the paper again, but guessing for now: we can’t look arbitrarily deep into objects to find these relations (I guess), but a limited search might be useful?

> I've tried that too, but it doesn't work with how subtyping knowledge is propagated:

One could then write `foo[A, B <: A]`, and you have to with Scalac, but I thought Dotty tried to accept code like yours? If you can call `foo`, then `A <:< B` already proves that `A <: B`, no need to propagate the constraint at all call sites.

Reply:
> > Note that this is not sufficient by any account.
>
> Good point, so I suspect I’m in favor of adding `Ev`. I think scalaz/cats have replacements for `<:<` (`Liskov`?), implemented with casts.

I just wanted to mention that [as of v2.13 of the standard library](https://github.com/scala/scala/pull/5623), `<:<` finally behaves like Liskov inequality, i.e. you can now do casts at almost no costs (they are the identity function) even if the related types are deeply nested in other type operators, using `substitute` of `liftCo`/`liftContra`. E.g. if you have evidence `p: A <:< B` you get `p.liftCo[List]: List[A] <: List[B]` for "free" (you still need to call `liftCo` manually).

Adding `Ev` to `<:<` seems appealing, but – as you discussed – has the drawback of working only in dotty. It's also potentially brittle since it uses inconsistent bounds. Since the `Ev` version is "extensional" (i.e. the type checker can reason about `Ev` directly), it might cover more cases, but I'm not sure about this. Maybe someone has an example where the `Ev` version works but the Liskov version (`liftCo`/`liftContra` + explicit cast) does not work?

Reply:
Thanks @sstucki, I had missed that. It's good to know!
 
It's true that `liftCo`/`liftContra` are sufficient in the `List` example, but they can become quite more cumbersome to use, as the user needs to pass in a type lambda when the coerced type does not have the straightforward one-argument shape of `T[_]`.

Here is an example where `liftCo`/`liftContra` actually cannot be used but where our approach works – or at least I don't see an easy way to use `liftCo`/`liftContra` in that case:

```scala
abstract class <:<[A, B] {
  type ConstrainedB >: A <: A & B // = B with constraints
  type ConstrainedA >: A | B <: B // = A with constraints
}
def foo[A,B<:A] = Set.empty[(A,B)]
def bar[A,B](implicit ev: B <:< A): Set[(A,B)] = foo[A,ev.ConstrainedB]
```

Not sure I understand why you say the bounds in `<:<` are inconsistent. Having an instance of it is a proof that they are indeed consistent. This might not happen in practice (for runtime performance reasons), but conceptually we would create a new instance of `A<:<B` for any `A` and `B` where we know `A<:B`, and encode that knowledge in the abstract type members.

EDIT: maybe you meant that they are not _syntactically_ consistent.

Reply:
You welcome @LPTK! Regarding your comments:

> Here is an example where `liftCo`/`liftContra` actually cannot be used (...)

Actually, this is a case for `liftContra`, but it's very cumbersome indeed. Here's a possible implementation:
```Scala
trait QFun[A, -X] { def apply[Y <: X]: Set[(A, Y)] }
def bar[A,B](implicit ev: B <:< A): Set[(A,B)] = {
  type Ctx[-X] = QFun[A, X]
  def qFoo[A] = new QFun[A, A] { def apply[Y <: A]: Set[(A, Y)] = foo }
  liftContra[B, A, Ctx](ev)(qFoo).apply
}
```
The `QFun` and `qFoo` machinery is necessary because polymorphism/universal quantification is restricted to trait and method definitions in Scala. If the language had System-F-style quantifiers, we could have written the example as follows:
```Scala
def bar[A,B](implicit ev: B <:< A): Set[(A,B)] = {
  type Ctx[-X] = ∀Y <: X. Set[(A, Y)]
  liftContra[B, A, Ctx](ev)(ΛY <: A.foo[A, Y])
}
```
But that's still much more clumsy than your version using bounds-based evidence directly.

> Not sure I understand why you say the bounds in `<:<` are inconsistent.

Sorry for the confusion re. consistency. I'm using the following terminology (adopted from my thesis): the bounds `A`, `B` of an abstract type `X >: A <: B` are
 * _consistent_ if we can prove that `A <: B` (in the context where `X` is defined),
 * _inconsistent_ if we cannot prove that `A <: B`,
 * _absurd_ if they are inconsistent even as closed types.

The difference between inconsistent and absurd bounds has to do with assumptions. E.g. the bounds of `X >: Any <: Nothing` are inconsistent, no matter the context, so we say they are absurd. But in your definition of `<:<`, we can neither prove that the bounds of `Ev >: A <: B` are consistent nor that they are absurd because we really don't know anything about `A` and `B` (they are abstract).

One could instead write "possibly inconsistent" but, personally, I prefer the shorter version.

As for the safety of inconsistent assumptions, note that, while inconsistency doesn't break safety of DOT, it [can make your compiler crash](https://github.com/lampepfl/dotty/issues/2887#issuecomment-316177441). In essence: reduction of open expressions under inconsistent bounds is unsafe, both for open terms _and_ open types. And contrary to open terms, open types _are_ routinely being reduced (at compile time).

Reply:
@sstucki
> The QFun and qFoo machinery is necessary because polymorphism/universal quantification is restricted to trait and method definitions in Scala

Interestingly, since the [recent introduction](https://github.com/lampepfl/dotty/pull/3464) of dependent function types, we now have a lightweight way to do first-class polymorphism! Indeed, we can write types such as `(x: {type T}) => List[x.T] => x.T`...

So, using it to define `liftContra` as in:

```scala
abstract class <:<[-From, +To] {
  protected type Ev >: From <: To
  def liftContra[F[-_]](f: F[To]): F[From] = f: F[Ev]
}
```

We can, today, write with Dotty:

```scala
def bar[A,B](implicit ev: B <:< A): Set[(A,B)] =
  ev.liftContra[
    [-X] => (p:{type T<:X}) => Set[(A,p.T)]
  ](p => foo[A,p.T])(new Poly{type T=B})
```

Where `trait Poly { type T }` is just used to avoid a type inference problem we run into with a bare `new{type T=B}`.